File: cpp/playfair.cpp
#include <iostream>
#include <string>
using namespace std;

char keyMatrix[5][5];
int posR[26], posC[26];   // store row and column of each letter

// Build matrix
void buildMatrix(string key) {
    bool used[26] = {false};
    used['J' - 'A'] = true; // J merges with I

    string k = "";
    for(char c : key) {
        if(isalpha(c)) {
            c = toupper(c);
            if(!used[c - 'A']) {
                used[c - 'A'] = true;
                k += c;
            }
        }
    }

    for(char c = 'A'; c <= 'Z'; c++)
        if(!used[c - 'A'])
            k += c;

    int idx = 0;
    for(int i = 0; i < 5; i++) {
        for(int j = 0; j < 5; j++) {
            keyMatrix[i][j] = k[idx];
            posR[k[idx] - 'A'] = i;
            posC[k[idx] - 'A'] = j;
            idx++;
        }
    }
}

// Prepare plaintext
string prepare(string text) {
    string p = "";
    for(char c : text) {
        if(isalpha(c)) {
            c = toupper(c);
            if(c == 'J') c = 'I';
            p += c;
        }
    }

    string res = "";
    for(int i = 0; i < p.size(); i++) {
        res += p[i];

        if(i + 1 < p.size() && p[i] == p[i+1])
            res += 'X';
    }

    if(res.size() % 2 == 1)
        res += 'X';

    return res;
}

// Encrypt a pair
string encPair(char a, char b) {
    int r1 = posR[a - 'A'], c1 = posC[a - 'A'];
    int r2 = posR[b - 'A'], c2 = posC[b - 'A'];

    if(r1 == r2) {  // same row
        c1 = (c1 + 1) % 5;
        c2 = (c2 + 1) % 5;
    }
    else if(c1 == c2) { // same column
        r1 = (r1 + 1) % 5;
        r2 = (r2 + 1) % 5;
    }
    else { // rectangle
        int temp = c1;
        c1 = c2;
        c2 = temp;
    }

    string res = "";
    res += keyMatrix[r1][c1];
    res += keyMatrix[r2][c2];
    return res;
}

// Decrypt a pair
string decPair(char a, char b) {
    int r1 = posR[a - 'A'], c1 = posC[a - 'A'];
    int r2 = posR[b - 'A'], c2 = posC[b - 'A'];

    if(r1 == r2) {   // same row
        c1 = (c1 + 4) % 5;
        c2 = (c2 + 4) % 5;
    }
    else if(c1 == c2) { // same column
        r1 = (r1 + 4) % 5;
        r2 = (r2 + 4) % 5;
    }
    else { // rectangle
        int temp = c1;
        c1 = c2;
        c2 = temp;
    }

    string res = "";
    res += keyMatrix[r1][c1];
    res += keyMatrix[r2][c2];
    return res;
}

string encrypt(string text) {
    string p = prepare(text);
    string out = "";

    for(int i = 0; i < p.size(); i += 2)
        out += encPair(p[i], p[i+1]);

    return out;
}

string decrypt(string text) {
    string out = "";

    for(int i = 0; i < text.size(); i += 2)
        out += decPair(text[i], text[i+1]);

    return out;
}

int main() {
    string key, text;

    cout << "Enter key: ";
    getline(cin, key);

    cout << "Enter text: ";
    getline(cin, text);

    buildMatrix(key);

    string enc = encrypt(text);
    string dec = decrypt(enc);

    cout << "\nEncrypted: " << enc << endl;
    cout << "Decrypted: " << dec << endl;

    return 0;
}

File: python/Play_Fair.py
def generate_key_matrix(keyword):
    keyword = keyword.upper().replace("J", "I") 
    matrix = []
    used = set()

    for char in keyword:
        if char.isalpha() and char not in used:
            used.add(char)
            matrix.append(char)

    # Add remaining alphabet
    for char in "ABCDEFGHIKLMNOPQRSTUVWXYZ": 
        if char not in used:
            used.add(char)
            matrix.append(char)

    # Convert to 5x5 matrix
    key_matrix = [matrix[i:i+5] for i in range(0, 25, 5)]
    return key_matrix

def prepare_message_pairs(message):
    message = message.upper().replace("J", "I")
    message = "".join([c for c in message if c.isalpha()])  # Keep letters only

    pairs = []
    i = 0
    while i < len(message):
        a = message[i]
        if i + 1 < len(message):
            b = message[i + 1]
            if a == b:  # Same letter → insert X
                pairs.append(a + "X")
                i += 1
            else:
                pairs.append(a + b)
                i += 2
        else:
            pairs.append(a + "X")  # Last single letter → add X
            i += 1
    return pairs

def find_position(matrix, char):
    for row in range(5):
        for col in range(5):
            if matrix[row][col] == char:
                return row, col
    return None

def encrypt_playfair(message, keyword):
    matrix = generate_key_matrix(keyword)
    pairs = prepare_message_pairs(message)
    encrypted = ""

    for a, b in pairs:
        row1, col1 = find_position(matrix, a)
        row2, col2 = find_position(matrix, b)

        if row1 == row2:  # Same row
            encrypted += matrix[row1][(col1 + 1) % 5]
            encrypted += matrix[row2][(col2 + 1) % 5]
        elif col1 == col2:  # Same column
            encrypted += matrix[(row1 + 1) % 5][col1]
            encrypted += matrix[(row2 + 1) % 5][col2]
        else:  # Rectangle rule
            encrypted += matrix[row1][col2]
            encrypted += matrix[row2][col1]

    return encrypted

def decrypt_playfair(ciphertext, keyword):
    matrix = generate_key_matrix(keyword)
    pairs = [ciphertext[i:i+2] for i in range(0, len(ciphertext), 2)]
    decrypted = ""

    for a, b in pairs:
        row1, col1 = find_position(matrix, a)
        row2, col2 = find_position(matrix, b)

        if row1 == row2:  # Same row → shift left
            decrypted += matrix[row1][(col1 - 1) % 5]
            decrypted += matrix[row2][(col2 - 1) % 5]
        elif col1 == col2:  # Same column → shift up
            decrypted += matrix[(row1 - 1) % 5][col1]
            decrypted += matrix[(row2 - 1) % 5][col2]
        else:  # Rectangle rule
            decrypted += matrix[row1][col2]
            decrypted += matrix[row2][col1]

    return decrypted


message = input("Enter plaintext: ")
keyword = input("Enter Key: ")

encrypted = encrypt_playfair(message, keyword)
decrypted = decrypt_playfair(encrypted, keyword)

print("\nKey Matrix:")
for row in generate_key_matrix(keyword):
    print(row)

print("Encrypted Text:", encrypted)
print("Decrypted Text:", decrypted)